src/arithmetic/stmt_simplify.cc

```diff
@@ -0,0 +1,167 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file stmt_simplify.cc
+ * \brief Statement simplifier based on analyzer
+ */
+#include <tvm/ir.h>
+#include <tvm/ir_pass.h>
+#include <tvm/arithmetic.h>
+#include <tvm/ir_mutator.h>
+#include <tvm/expr_operator.h>
+#include <tvm/arithmetic.h>
+#include "arithmetic/Simplify.h"
+
+namespace tvm {
+namespace arith {
+// statement simplifier
+using namespace ir;
+
+class StmtSimplifier : public IRMutator {
+ public:
+  Stmt Mutate_(const For* op, const Stmt& s) final {
+    Var loop_var(op->loop_var.node_);
+    analyzer_.Bind(loop_var, Range::make_by_min_extent(op->min, op->extent));
+    return IRMutator::Mutate_(op, s);
+  }
+
+  // IfThenElse
+  Stmt Mutate_(const IfThenElse* op, const Stmt& s) {
+    Expr condition = this->Mutate(op->condition);
+    Stmt then_case, else_case;
+    {
+      ConstraintContext ctx(&analyzer_, condition);
+      then_case = this->Mutate(op->then_case);
+    }
+    if (op->else_case.defined()) {
+      ConstraintContext ctx(&analyzer_, Mutate(Not::make(condition)));
+      else_case = this->Mutate(op->else_case);
+    }
+    if (is_one(condition)) return op->then_case;
```

Comment:
Shouldn't we return `then_case` instead of `op->then_case` here? (And same thing with the else_case)